examples/hybrid_jobs/6_QNSPSA_optimizer_with_embedded_simulator/qnspsa_with_embedded_simulator.ipynb

```diff
@@ -0,0 +1,1360 @@
+    "ax.legend(['GD', 'QNG', 'QN-SPSA', 'SPSA'])#, 'ground energy'])\n",
+    "ax.set_xlabel(\"Step\")\n",
+    "ax.set_ylabel(\"Loss\")\n",
+    "plt.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "83a800ae",
+   "metadata": {},
+   "source": [
+    "The above plot shows how the loss function drops over optimization steps. The purple/green solid line represents the average of the `spsa_repeats` number of curves for SPSA/QN-SPSA. The green shaded area stands for the envelope of all `spsa_repeats` curves for QN-SPSA. The envelope of the SPSA is not plotted since it is too noisy to visualize. \n",
+    "\n",
+    "SPSA turns out to provide really noisy results for such a use case. On the other hand, GD, QNG and QN-SPSA are able to minimize the loss function to be below -13, giving a estimated ground state of -14 for the max cut problem. I would like to understand how good they are doing finding the global minimum. Luckily for a 10-node graph, it is possible to loop through all possibilities to find the analytical ground state to the max cut problem. In this example, the ground state engergy is -16, as given by the following paragraph. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "id": "ddf4f323",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "The theoretical ground state of the max cut problem is -16\n"
+     ]
+    }
+   ],
+   "source": [
+    "nodes = n_qubits = 10\n",
+    "edges = 20\n",
+    "seed = 197\n",
+    "\n",
+    "g = nx.gnm_random_graph(nodes, edges, seed=seed)\n",
+    "\n",
+    "def powerset(iterable):\n",
+    "    s = list(iterable)\n",
+    "    return chain.from_iterable(combinations(s, r) for r in range(len(s)+1))\n",
+    "\n",
+    "max_cut = 0\n",
+    "for S in powerset(range(nodes)): \n",
+    "    temp_cut = nx.cut_size(g, S)\n",
+    "    if max_cut < temp_cut:\n",
+    "        max_cut = temp_cut\n",
+    "print(\"The theoretical ground state of the max cut problem is\", -max_cut)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "6c858417",
+   "metadata": {},
+   "source": [
+    "Recall that QAOA only provides an estimated result to the ground state (that's why it is named \"Approximate\"), the GD, QNG and QN-SPSA perform reasonably well minimizing the loss function. Yet I will not draw a simple conclusion here on which optimizer is better finding the global minimum. The max cut problem gives a non-convex and complex landscape, and the optimization result highly depends on the graph structure and the initialization (Feel free to rerun the experiment above with a different seed to see this effect). In my different experiment trials with different graph structures, I notice that GD and QNG sometimes get stuck with local minima and fail to further optimize, while QN-SPSA appears to be more consistent in minimizing the loss function. "
```

Comment:
"estimated result to the ground state": Let's be more specific: QAOA being a variational algorithm, it provides an upper bound.